### ISSUE TITLE
[Linux] X509 chains with user principal name (UPN) name constraints fail to build

### ISSUE THREAD

**Issue by skeiffer** [dotnet/runtime#87413]
### Description

X509Chain.Build() throws a `CryptographicException` when building a chain that contains a user principal name (UPN) name constraint. This prevents the caller from getting any information about the validity of the chain.

This problem is due to openSSL not supporting UPN based name constraints and `MapVerifyErrorToChainStatus` not mapping the status code that openSSL provides for this error (51).

This only seems affect systems that rely on OpenSSL. 
* Windows supports UPN name constraints correctly.
* MacOS does not support UPN name constraints but skips the check.

### Reproduction Steps

Add the following unit test to `./src/libraries/System.Security.Cryptography/tests/X509Certificates/DynamicChainTests.cs`

``` C#
[Fact]
[SkipOnPlatform(TestPlatforms.Windows, "Windows supports UPN name constraints.")]
[SkipOnPlatform(PlatformSupport.AppleCrypto, "macOS ignores UPN name constraints.")]
public static void UnknownNameConstraintViolation_UserPrincipalName()
{
    SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
    builder.AddUserPrincipalName("johndoe@microsoft.com");

    // exlude UPN name constraint for @microsoft.com
    string nameConstraints = "3024A1223020A01E060A2B060104018237140203A0100C0E406D6963726F736F66742E636F6D";

    TestNameConstrainedChain(nameConstraints, builder, (bool result, X509Chain chain) => {
        Assert.False(result, "chain.Build");
        Assert.Equal(PlatformNameConstraints(X509ChainStatusFlags.HasNotSupportedNameConstraint), chain.AllStatusFlags());
    });
}
```

### Expected behavior

`X509Chain.Build()` does not throw an exception but adds a chain status flag to the chain. (ex: `X509ChainStatusFlags.HasNotSupportedNameConstraint` or similar)

### Actual behavior

`X509Chain.Build()` throws a `CryptographicException`

**Comment by bartonjs** [comment]
Assuming that Windows handles UPN constraints, there are really two issues here:

1) We should add UPN constraint enforcement to our secondary checks layer for macOS and Linux.
2) We should make up a test with an actually unhandled constraint and see what Windows does for the rest of chain processing (doing/skipping revocation checks, aborting the chain build, etc... as well as what all codes get mapped in).

If Windows actually doesn't do the UPN constraint enforcement (and it's just an app decision) then part 1 goes away.

**Comment by Jozkee** [comment]
Assuming @vcsjones is not working on this already. @skeiffer if you want to invest in fixing this feel free to do so considering @bartonjs concerns.

**Comment by vcsjones** [comment]
Yep, sorry - I might be able to look at this later but not in the short term.

**Comment by skeiffer** [comment]
Hi @Jozkee. I do not have the time to perform a more through implementation that @bartonjs is asking for. If this work could be assigned it would be greatly appreciated. If internal business justification is needed for scheduling/planning, we can provide that.

**Comment by WyattHoutz** [comment]
Hello @jeffhandley, @bartonjs , @jozkee, I see this is set to "Future" milestone. Will this be fixed in any upcoming release? We are still waiting on a fix. (fyi @skeiffer )

**Comment by bartonjs** [comment]
It's not "not" being fixed, but it probably fell into a crack and got overlooked.  "Future" doesn't mean "no", it just means that we won't say .NET 9/10/11/etc "isn't done" until this is fixed.  (We're not always good about consistently applying that with the milestone assignment, but it's supposed to mean "committed"... and the best way to be "committed" is to be done.)

**Comment by vcsjones** [comment]
I can take a look.

**Comment by vcsjones** [comment]
I thought @bartonjs left a comment here, but: it will be addressed in .NET 10 for Linux.

**Comment by bartonjs** [comment]
> I thought @bartonjs left a comment here

I believe that was on an alternate communication channel, like email.  Yay multi-modal communication.

**Comment by vcsjones** [comment]
> any chance you could share what was found

To quickly summarize:

1. Windows is the only platform that handles name constraints of type `otherName`. Other platforms are going to fail to build a chain by returning false, and having either `InvalidNameConstraints` or ``HasNotSupportedNameConstraint` as a chain status. To my knowledge, there is no specification of how `otherName` is handled with name constraints. From [RFC 5280](https://www.rfc-editor.org/rfc/rfc5280.html#section-4.2.1.10)

    > The syntax and semantics for name constraints for otherName,
   ediPartyName, and registeredID are not defined by this specification,
   however, syntax and semantics for name constraints for other name
   forms may be specified in other documents. 

    I was unable to find any other RFCs that give semantics on processing certificate chains with `otherName`. This is likely why non-Windows platforms simply don't implement `otherName` handling of name constraints.

 > MacOS does not support UPN name constraints but skips the check.

    macOS (and other Apple platforms) will report `InvalidNameConstraints` as of macOS 15.4 and iOS-like platforms as of 18.4.

**Comment by vcsjones** [comment]
> Will it be released in the next .NET 10 preview?

No, the next preview is Preview 4 but this missed the cut off for that. The preview it will be available in is .NET 10 Preview 5.
